File: scripts/boolean-shapely.js
const PythonShell = require('python-shell');

/**
 * Boolean Shapely helper script
 *
 * @param {string} operation ("crosses", "contains")
 * @param {Geometry|Feature<any>} feature1 GeoJSON Feature
 * @param {Geometry|Feature<any>} feature2 GeoJSON Feature
 * @returns {Promise<Boolean>} true/false
 * @example
 * const operation = 'crosses';
 * const feature1 = lineString([[-2, 2], [1, 1]]);
 * const feature2 = lineString([[1, 1], [1, 2], [1, 3], [1, 4]]);
 *
 * booleanShapely(operation, feature1, feature2)
 *     .then(result => console.log(result));
 */
module.exports = function (operation, feature1, feature2) {
    // Convert Feature to Geometry
    if (feature1.geometry) feature1 = feature1.geometry;
    if (feature2.geometry) feature2 = feature2.geometry;

    const options = {
        scriptPath: __dirname,
        args: [
            operation,
            JSON.stringify(feature1),
            JSON.stringify(feature2)
        ]
    };
    return new Promise((resolve, reject) => {
        const pyshell = new PythonShell('boolean-shapely.py', options);
        pyshell.on('message', message => {
            return resolve(message === 'True');
        });
        pyshell.end(error => {
            if (error) throw error;
            return reject(error);
        });
    });
};

File: scripts/boolean-shapely.py
#!/usr/bin/env python
"""Shapely helper script for Boolean validation"""

import sys
import json
from shapely.geometry import asShape

def shapely(operation, geometry1, geometry2):
    geometry1 = asShape(json.loads(geometry1))
    geometry2 = asShape(json.loads(geometry2))

    if operation == 'crosses':
        print geometry1.crosses(geometry2)
    elif operation == 'contains':
        print geometry1.conains(geometry2)

if __name__ == '__main__':
    OPERATION = sys.argv[1]
    FEATURE1 = sys.argv[2]
    FEATURE2 = sys.argv[3]
    shapely(OPERATION, FEATURE1, FEATURE2)
